lib/pbench/server/api/resources/upload_api.py

```diff
@@ -271,13 +287,13 @@ def put(self, filename_arg):
                 try:
                     os.remove(md5_full_path)
```

Comment:
If we die trying to create the link file at line 285 we carefully try to remove the hash file, but we don't touch the uploaded data file or the `Dataset` DB entry.  It seems like we should be using a more structured approach here with nested operations which can be backed out in the event of problems.

Reply:
This is a real problem, thanks for catching that.  I'll work in as another commit.

Reply:
Actually, not sure this is a problem.

The `with` statement _always_ removes the uploaded data file using its "named" temporary entry.  That removal will only delete the file on disk if there is no second link to it.  So once the `os.link()` is successful, the file entry is in its final resting place.

The Dataset is always a permanent record, it is never removed.  So it will live in a partially uploaded state.

Reply:
Tracked in #2299.

Reply:
I'm fine with deferring the task of restoring invariants after an error; however, I was hoping that deferring the creation of the `Dataset` until after the upload (which would avoid much of the need for that restoration) would be simple enough to be included in this PR.  Is there some aspect of that which I'm missing?

Reply:
Filed as a separate commit as I am not convinced of this change just yet.

I have implemented what you requested above.

I think we want to record the `Dataset` in such a way as to know details of on going uploads (potentially updates the bytes received dynamically) so that users can see their uploads in process.  And so that they can see the failures from the UI.

We just need a way to know when it is okay for a new upload to over-write a previously failed one.

Reply:
> I think we want to record the `Dataset` in such a way as to know details of on going uploads (potentially updates the bytes received dynamically) so that users can see their uploads in process. And so that they can see the failures from the UI.

Yes, that was the idea; we start in UPLOADING and go to UPLOADED at the end of the upload. Now, there's nothing right now in the UI that's sensitive to the `Dataset` state ... but the whole state tracker was designed to allow us to add some smarts in that area. And, yes, it means we need more sophisticated error recovery than simply deferring everything to the end.

> We just need a way to know when it is okay for a new upload to over-write a previously failed one.

Indeed.

Reply:
Owing to the fact that I was only reviewing the deltas, I somehow missed the fact that we update the `Dataset` state after the upload completes...so, yes, my proposed change is _not_ a good idea, and Pete can (and did) roll it back.  (Vely clever of you, Peter, to keep it as a separate commit!)

I really thought that there wasn't any more code at the end of that function....  :-(